File: cpp/binary_tree/AVL_Tree.cpp
#include<bits/stdc++.h> 
using namespace std; 
  
// An AVL tree
class AVLtree  
{
	//Node of AVLtree
	struct Node
	{
		int key;  
    	Node *left;  
    	Node *right;
		int height;
	};
	
	//Parent node i.e root of AVLtree
	Node *root;
	
	// A utility function to get height of the tree  
	int height(Node *N)  
	{  
	    if (N == NULL) return 0;  
	    return N->height;  
	}  
	
	/*Helper function that dynamically allocates
	a new node with the given key and  NULL left 
	and right pointers*/
	Node* newNode(int key)  
	{  
	    Node* node = new Node(); 
	    node->key = key;  
	    node->left = NULL;  
	    node->right = NULL;  
	    node->height = 1;
	    return(node);  
	}  
	  
	/* A utility function to right 
	rotate subtree rooted with y*/  
	Node *rightRotate(Node *y)  
	{  
	    Node *x = y->left;  
	    Node *T2 = x->right;  
	    // Perform rotation  
	    x->right = y;  
	    y->left = T2;  
	    // Update heights  
	    y->height = max(height(y->left),height(y->right)) + 1;  
	    x->height = max(height(x->left),height(x->right)) + 1;
	    // Return new root  
	    return x;  
	}  
	  
	/* A utility function to left  
	rotate subtree rooted with x*/
	Node *leftRotate(Node *x)  
	{  
	    Node *y = x->right;  
	    Node *T2 = y->left;  
	    // Perform rotation  
	    y->left = x;  
	    x->right = T2;  
	    // Update heights  
	    x->height = max(height(x->left),height(x->right)) + 1;  
	    y->height = max(height(y->left),height(y->right)) + 1;
	    // Return new root  
	    return y;  
	}  
	  
	// Get Balance factor of node N  
	int getBalance(Node *N)  
	{  
	    if (N == NULL) return 0;  
	    return height(N->left) - height(N->right);  
	}  
	  
	Node* _insert(Node* node, int key)  
	{  
	    //1. Perform the normal BST rotation
	    if (node == NULL)
		{
			return(newNode(key));
		}   
	    if (key < node->key)
		{
			node->left = _insert(node->left, key);
		}   
	    else if (key > node->key) 
		{
			node->right = _insert(node->right, key);
		}  
	    else//Equal keys not allowed  
	    {
			return node;
		}      
	  
	    //2. Update height of this ancestor node
	    node->height = 1 + max(height(node->left),height(node->right));  
	  
	    /*3. Get the balance factor of this  
	    ancestor node to check whether  
	    this node became unbalanced */
	    int balance = getBalance(node);  
	  
		// If this node becomes unbalanced, 
	    // then there are 4 cases  
	  
	    // Left Left Case  
	    if (balance > 1 && key < node->left->key) 
		{
			return rightRotate(node);
		}
	    // Right Right Case  
	    if (balance < -1 && key > node->right->key) 
		{
			return leftRotate(node);
		}
	    // Left Right Case  
	    if (balance > 1 && key > node->left->key)  
	    {  
	        node->left = leftRotate(node->left);  
	        return rightRotate(node);  
	    }
	    // Right Left Case  
	    if (balance < -1 && key < node->right->key)  
	    {  
	        node->right = rightRotate(node->right);  
	        return leftRotate(node);  
	    }
	    /* return the (unchanged) node pointer */
	    return node;  
	}  
	  
	/* Given a non-empty binary search tree,  
	return the node with minimum key value  
	found in that tree. Note that the entire  
	tree does not need to be searched. */
	Node * minValueNode(Node* node)  
	{  
	    Node* current = node;
	    // loop down to find the leftmost leaf
	    while (current->left != NULL)
		{
			current = current->left;
		}
	    return current;  
	}  
	  
	/* Recursive function to delete a node  
	with given key from subtree with  
	given root. It returns root of the  
	modified subtree*/ 
	Node* _deleteNode(Node* root, int key)  
	{  
	      
	    // STEP 1: PERFORM STANDARD BST DELETE  
	    if (root == NULL) return root;  
	    /* If the key to be deleted is smaller  
	    than the root's key, then it lies 
	    in left subtree*/
	    if ( key < root->key ) root->left = _deleteNode(root->left, key);  
	    /* If the key to be deleted is greater  
	    than the root's key, then it lies  
	    in right subtree*/  
	    else if( key > root->key ) root->right = _deleteNode(root->right, key);  
	    /*if key is same as root's key, then  
	    This is the node to be deleted */
	    else
	    {  
	        // node with only one child or no child  
	        if( (root->left == NULL) || (root->right == NULL) )  
	        {  
	            Node *temp = root->left ?  root->left : root->right;  
	            // No child case  
	            if (temp == NULL)  
	            {  
	                temp = root;  
	                root = NULL;  
	            }  
	            else// One child case
				{
					*root = *temp;
				}
	            free(temp);  
	        }  
	        else
	        {  
	            /* node with two children: Get the inorder  
	            successor (smallest in the right subtree)*/
	            Node* temp = minValueNode(root->right);  
	            /* Copy the inorder successor's  
	            data to this node*/ 
	            root->key = temp->key;  
	            // Delete the inorder successor  
	            root->right = _deleteNode(root->right,temp->key);  
	        }  
	    }
	    // If the tree had only one node then return  
	    if (root == NULL) return root;  
	  
	    // STEP 2: UPDATE HEIGHT OF THE CURRENT NODE  
	    root->height = 1 + max(height(root->left),height(root->right));  
	  
	    /* STEP 3: GET THE BALANCE FACTOR OF  
	    THIS NODE (to check whether this  
	    node became unbalanced)*/
	    int balance = getBalance(root);
	    
	    // If this node becomes unbalanced,  
	    // then there are 4 cases  
	  
	    // Left Left Case  
	    if (balance > 1 && getBalance(root->left) >= 0)
		{
			return rightRotate(root);
		}
	    // Left Right Case  
	    if (balance > 1 && getBalance(root->left) < 0)  
	    {  
	        root->left = leftRotate(root->left);  
	        return rightRotate(root);  
	    } 
	    // Right Right Case  
	    if (balance < -1 && getBalance(root->right) <= 0) 
		{
			return leftRotate(root);
		}
	    // Right Left Case  
	    if (balance < -1 && getBalance(root->right) > 0)  
	    {  
	        root->right = rightRotate(root->right);  
	        return leftRotate(root);  
	    }
	    return root;  
	}  
	  
	/*A utility function to print preorder  
	traversal of the tree.The function 
	also prints height of every node */
	void preOrder(Node *root)  
	{  
	    if(root != NULL)  
	    {  
	        cout << root->key << " ";  
	        preOrder(root->left);  
	        preOrder(root->right);  
	    }  
	}  
	
public:
	AVLtree()
	{
		root=NULL;
	}	
	
	void insert(int x)
	{
		root=_insert(root,x);
	}
	
	void deleteNode(int x)
	{
		root=_deleteNode(root,x);
	}
	
	void display() 
	{
		preOrder(root);
        cout << "\n";
    }
};  



int main()  
{
	AVLtree t;  

    t.insert(9);  
    t.insert(5);  
    t.insert(10);  
    t.insert(0);  
    t.insert(6);  
    t.insert(11);  
    t.insert(-1);  
    t.insert(1);  
    t.insert(2);
  
    cout << "Preorder traversal of the constructed AVL tree is \n";  
    t.display();
    t.deleteNode(10);

    cout << "\nPreorder traversal after deletion of 10 \n";  
    t.display();  
    return 0;  
}  

File: python/tree/AVL_Tree.py
import sys

class Node:
	def  __init__(self, data):
		self.data = data
		self.parent = None
		self.left = None
		self.right = None
		self.bf = 0

class AVLTree:

	def __init__(self):
		self.root = None

	def __printHelper(self, currPtr, indent, last):
		# print the tree structure on the screen
	   	if currPtr != None:
			sys.stdout.write(indent)
			if last:
			  	sys.stdout.write("R----")
			  	indent += "     "
			else:
				sys.stdout.write("L----")
				indent += "|    "

			print currPtr.data

			self.__printHelper(currPtr.left, indent, False)
			self.__printHelper(currPtr.right, indent, True)
	
	def __searchTreeHelper(self, node, key):
		if node == None or key == node.data:
			return node

		if key < node.data:
			return self.__searchTreeHelper(node.left, key)
		return self.__searchTreeHelper(node.right, key)

	def __deleteNodeHelper(self, node, key):
		# search the key
		if node == None: 
			return node
		elif key < node.data:
			node.left = self.__deleteNodeHelper(node.left, key)
		elif key > node.data: 
			node.right = self.__deleteNodeHelper(node.right, key)
		else:
			# the key has been found, now delete it

			# case 1: node is a leaf node
			if node.left == None and node.right == None:
				node = None

			# case 2: node has only one child
			elif node.left == None:
				temp = node
				node = node.right

			elif node.right == None:
				temp = node
				node = node.left

			# case 3: has both children
			else:
				temp = minimum(node.right)
				node.data = temp.data
				node.right = self.__deleteNodeHelper(node.right, temp.data)

			# Write the update balance logic here 
			# YOUR CODE HERE
		return node

	# update the balance factor the node
	def __updateBalance(self, node):
		if node.bf < -1 or node.bf > 1:
			self.__rebalance(node)
			return;

		if node.parent != None:
			if node == node.parent.left:
				node.parent.bf -= 1

			if node == node.parent.right:
				node.parent.bf += 1

			if node.parent.bf != 0:
				self.__updateBalance(node.parent)

	 # rebalance the tree
	def __rebalance(self, node):
		if node.bf > 0:
			if node.right.bf < 0:
				self.rightRotate(node.right)
				self.leftRotate(node)
			else:
				self.leftRotate(node)
		elif node.bf < 0:
			if node.left.bf > 0:
				self.leftRotate(node.left)
				self.rightRotate(node)
			else:
				rightRotate(node)

	def __preOrderHelper(self, node):
		if node != None:
			sys.stdout.write(node.data + " ")
			self.__preOrderHelper(node.left)
			self.__preOrderHelper(node.right)

	def __inOrderHelper(self, node):
		if node != None:
			self.__inOrderHelper(node.left)
			sys.stdout.write(node.data + " ")
			self.__inOrderHelper(node.right)

	def __postOrderHelper(self, node):
		if node != None:
			self.__postOrderHelper(node.left)
			self.__postOrderHelper(node.right)
			std.out.write(node.data + " ")

	# Pre-Order traversal
	# Node->Left Subtree->Right Subtree
	def preorder(self):
		self.__preOrderHelper(self.root)

	# In-Order traversal
	# Left Subtree -> Node -> Right Subtree
	def __inorder(self):
		self.__inOrderHelper(self.root)

	# Post-Order traversal
	# Left Subtree -> Right Subtree -> Node
	def __postorder(self):
		self.__postOrderHelper(self.root)

	# search the tree for the key k
	# and return the corresponding node
	def searchTree(self, k):
		return self.__searchTreeHelper(self.root, k)

	# find the node with the minimum key
	def minimum(self, node):
		while node.left != None:
			node = node.left
		return node

	# find the node with the maximum key
	def maximum(self, node):
		while node.right != None:
			node = node.right
		return node

	# find the successor of a given node
	def successor(self, x):
		# if the right subtree is not null,
		# the successor is the leftmost node in the
		# right subtree
		if x.right != None:
			return self.minimum(x.right)

		# else it is the lowest ancestor of x whose
		# left child is also an ancestor of x.
		y = x.parent
		while y != None and x == y.right:
			x = y
			y = y.parent
		return y

	# find the predecessor of a given node
	def predecessor(self, x):
		# if the left subtree is not null,
		# the predecessor is the rightmost node in the 
		# left subtree
		if x.left != None:
			return self.maximum(x.left)

		y = x.parent
		while y != None and x == y.left:
			x = y
			y = y.parent
		return y

	# rotate left at node x
	def leftRotate(self, x):
		y = x.right
		x.right = y.left
		if y.left != None:
			y.left.parent = x

		y.parent = x.parent;
		if x.parent == None:
			self.root = y
		elif x == x.parent.left:
			x.parent.left = y
		else:
			x.parent.right = y
		y.left = x
		x.parent = y

		# update the balance factor
		x.bf = x.bf - 1 - max(0, y.bf)
		y.bf = y.bf - 1 + min(0, x.bf)

	# rotate right at node x
	def rightRotate(self, x):
		y = x.left
		x.left = y.right;
		if y.right != None:
			y.right.parent = x
		
		y.parent = x.parent;
		if x.parent == None:
			self.root = y
		elif x == x.parent.right:
			x.parent.right = y
		else:
			x.parent.left = y
		
		y.right = x
		x.parent = y

		# update the balance factor
		x.bf = x.bf + 1 - min(0, y.bf)
		y.bf = y.bf + 1 + max(0, x.bf)

	# insert the key to the tree in its appropriate position
	def insert(self, key):
		# PART 1: Ordinary BST insert
		node =  Node(key)
		y = None
		x = self.root

		while x != None:
			y = x
			if node.data < x.data:
				x = x.left
			else:
				x = x.right

		# y is parent of x
		node.parent = y
		if y == None:
			self.root = node
		elif node.data < y.data:
			y.left = node
		else:
			y.right = node

		# PART 2: re-balance the node if necessary
		self.__updateBalance(node)


	# delete the node from the tree
	def deleteNode(self, data):
		return self.__deleteNodeHelper(self.root, data)

	# print the tree structure on the screen
	def prettyPrint(self):
		self.__printHelper(self.root, "", True)

if __name__ == '__main__':
	bst = AVLTree()
	bst.insert(1)
	bst.insert(2)
	bst.insert(3)
	bst.insert(4)
	bst.insert(5)
	bst.insert(6)
	bst.insert(7)
	bst.insert(8)
	bst.prettyPrint()